src/sdk/namespace_s3.js

```
@@ -139,13 +139,13 @@ class NamespaceS3 {
             // If set, part_number is positive integer from 1 to 10000.
             // Usually part number is not provided and then we read a small "inline" range
             // to reduce the double latency for small objects.
-            if (!params.part_number) {
+            if (!params.part_number && !request.Key.endsWith('/')) {
```

why is this the condition? why are keys ending with / should be different?
if needed, define this condition in a const boolean to avoid repeating it.

---

getObject when object is directory is failing because of invalid range, and I think that we can use headObject in this case instead.
WDYT?

---

what puzzles me is why getObject(0-4k) for dir fails? Does getObject(0-4k) for any 0 size object fail on invalid range like that too? or is it special for dirs?

---

@guymguym it's on 0 size objects too

---

oh that's not good.

---

We need to fix for 0 size objects too, we can't let that fail of course.
And for object sizes between 1-4095 (not 0) does it works ok although we request more than the object size? and only for 0 it doesn't? where did you test it? AWS?

---

yes, checked it on AWS.
and I tried it also with object of size 918 bytes and getObject works with range so this scenario is ok.

---

got it. any suggestions how to handle 0 size objects?

---

what about catch that error and try headObject instead?

---

can work... but perhaps it would be good to separate this whole change to another PR?
It's getting more complex to piggyback this PR.

---

it's a part of the spark fix (the spark session object that the reporter uses does headObject as part of its process)